src/Data/Text/Internal/Reverse.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnliftedFFITypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+{-# OPTIONS_HADDOCK not-home #-}
+
+-- | Implements 'reverse', using efficient C routines by default.
+module Data.Text.Internal.Reverse (reverse) where
+
+#if !defined(PURE_HASKELL)
+import GHC.Exts as Exts
+import Control.Monad.ST.Unsafe (unsafeIOToST)
+import Foreign.C.Types (CSize(..))
+#else
+import Control.Monad.ST (ST)
+import Data.Text.Internal.Encoding.Utf8 (utf8LengthByLeader)
+#endif
+import Prelude hiding (reverse)
+import Data.Text.Internal (Text(..))
+import Control.Monad.ST (runST)
+import qualified Data.Text.Array as A
+
+-- | /O(n)/ Reverse the characters of a string.
+--
+-- Example:
+--
+-- $setup
+-- >>> T.reverse "desrever"
+-- "reversed"
+reverse ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  Text -> Text
+#if defined(PURE_HASKELL)
+reverse (Text ba off len) = runST $ do
+    dest <- A.new len
+    -- let bytes = A.toList ba off len
+    -- _ <- reversePoints bytes (len - 1) dest
+    _ <- reversePoints' ba off dest (len - 1)
+    result <- A.unsafeFreeze dest
+    pure $ Text result 0 len
+
+reversePoints' :: A.Array -> Int -> A.MArray s -> Int -> ST s ()
+reversePoints' ba p_in dest p_out =
+    let pointLength = utf8LengthByLeader (A.unsafeIndex ba p_in)
+    in do
+        A.copyI pointLength dest (p_out - pointLength + 1) ba p_in
+        if p_out - pointLength >= 0
+            then reversePoints' ba (p_in + pointLength) dest (p_out - pointLength)
+            else pure ()
```

Comment:
```
All
  Pure
    tiny
      reverse
        Text:     OK (0.29s)
          32.7 ns ± 1.3 ns,       same as baseline
        LazyText: OK (0.17s)
          40.9 ns ± 2.7 ns, 10% less than baseline
    ascii-small
      reverse
        Text:     OK (0.18s)
          370  μs ±  24 μs,       same as baseline
        LazyText: OK (0.19s)
          367  μs ±  22 μs,       same as baseline
    ascii
      reverse
        Text:     OK (1.01s)
          308  ms ±  18 ms,       same as baseline
        LazyText: OK (0.98s)
          303  ms ± 3.0 ms,  1% more than baseline
    english
      reverse
        Text:     OK (0.18s)
          21.3 ms ± 1.4 ms,       same as baseline
        LazyText: OK (0.14s)
          20.2 ms ± 1.3 ms,       same as baseline
    russian
      reverse
        Text:     OK (0.15s)
          35.0 μs ± 2.9 μs, 14% less than baseline
        LazyText: OK (0.15s)
          34.9 μs ± 2.9 μs, 14% less than baseline
    japanese
      reverse
        Text:     OK (0.88s)
          25.8 μs ± 1.5 μs,       same as baseline
        LazyText: OK (0.88s)
          25.8 μs ± 1.6 μs,       same as baseline

All 12 tests passed (6.01s)
```

Indeed a lot faster for Russian, which I assume uses >1byte points. I would assume it would be faster for Japanese, too, but maybe the benchmarks themselves are wonky.